lib/dangermattic/plugins/android_unit_test_pr_checker.rb

```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+require_relative 'utils/git_utils'
+
+module Danger
+  # Plugin to detect classes without Unit Tests in a PR.
+  class AndroidUnitTestPRChecker < Plugin
+    ANY_CLASS_DETECTOR = /class ([A-Z]\w+)\s*(.*?)\s*{/
+    NON_PRIVATE_CLASS_DETECTOR = /(?:\s|public|internal|protected|final|abstract|static)*class ([A-Z]\w+)\s*(.*?)\s*{/
+
+    CLASSES_EXCEPTIONS = [
+      /ViewHolder$/,
+      /Module$/
+    ].freeze
+
+    SUBCLASSES_EXCEPTIONS = [
+      /(Fragment|Activity)\b/,
+      /RecyclerView/
+    ].freeze
+
+    UNIT_TESTS_BYPASS_PR_LABEL = 'unit-tests-exemption'
+
+    # Check and warns about missing unit tests for a Git diff, with optional classes/subclasses to ignore and an
+    # optional PR label to bypass the checks.
+    #
+    # @param classes_exceptions [Array<String>] Optional list of regexes matching class names to exclude from the
+    # check.
+    #   Defaults to CLASSES_EXCEPTIONS.
+    # @param subclasses_exceptions [Array<String>] Optional list of regexes matching base class names to exclude from
+    # the check.
+    #   Defaults to SUBCLASSES_EXCEPTIONS.
+    # @param bypass_label [String] Optional label to indicate we can bypass the check. Defaults to
+    #   UNIT_TESTS_BYPASS_PR_LABEL.
+    # @return [void]
+    #
+    # @example Check missing unit tests
+    #   check_missing_tests()
+    #
+    # @example Check missing unit tests excluding certain classes and subclasses
+    #   check_missing_tests(classes_exceptions: [/ViewHolder$/], subclasses_exceptions: [/RecyclerView/])
+    #
+    # @example Check missing unit tests with a custom bypass label
+    #   check_missing_tests(bypass_label: 'BypassTestCheck')
+    def check_missing_tests(classes_exceptions: CLASSES_EXCEPTIONS, subclasses_exceptions: SUBCLASSES_EXCEPTIONS,
+                            bypass_label: UNIT_TESTS_BYPASS_PR_LABEL)
+      list = find_classes_missing_tests(
+        git_diff: git.diff,
+        classes_exceptions: classes_exceptions,
+        subclasses_exceptions: subclasses_exceptions
+      )
+
+      return if list.empty?
+
+      if danger.github.pr_labels.include?(bypass_label)
+        list.each do |c|
+          warn("Class `#{c.classname}` is missing tests, but `#{bypass_label}` label was set to ignore this.")
+        end
+      else
+        list.each do |c|
+          failure("Please add tests for class `#{c.classname}` (or add `#{bypass_label}` label to ignore this).")
+        end
+      end
+    end
+
+    private
+
+    ClassViolation = Struct.new(:classname, :file)
+
+    # @param [Git::Diff] git_diff the object
+    # @return [Array<ClassViolation>] An array of `ClassViolation` objects for each added class that is missing a test
+    def find_classes_missing_tests(git_diff:, classes_exceptions:, subclasses_exceptions:)
```

Comment:
Indeed, this will be part of the code shared between the iOS and the Android plugins 👍 and I guess we'll also be able to reuse the test scenarios as test examples as well.